### PR TITLE
[BEAM-1797] add CoGroupByKey to chapter 'Using GroupByKey'  

### PR DESCRIPTION
**Pull request by mingmxu**
`CoGroupByKey` is the main component to create a JOIN operation in Beam pipeline, add the usage to chapter `Using GroupByKey`.

### CODE REVIEW

**Comment by davorbonaci** [comment]
R: @melap

File: src/documentation/programming-guide.md [hunk 1]
```diff
@@ -464,8 +464,33 @@ tree, [2]
 
 Thus, `GroupByKey` represents a transform from a multimap (multiple keys to individual values) to a uni-map (unique keys to collections of values).
 
-> **A Note on Key/Value Pairs:** Beam represents key/value pairs slightly differently depending on the language and SDK you're using. In the Beam SDK for Java, you represent a key/value pair with an object of type `KV<K, V>`. In Python, you represent key/value pairs with 2-tuples.
+`CoGroupByKey` is a `PTransform` based on `GroupByKey`, differs from `GroupByKey`, it merges multiple key/value `PCollection`s with the same key type, and emits a collection of `KV<K, CoGbkResult>` pairs. It's usually used as a JOIN operation in Beam pipeline(describe in [design-your-pipeline#multiple-sources]({{ site.baseurl }}/documentation/pipelines/design-your-pipeline/#multiple-sources)). 
+
+Given the input collections as below:
```
Comment (melap) [inline src/documentation/programming-guide.md]: remove "as"

File: src/documentation/programming-guide.md [hunk 2]
```diff
@@ -464,8 +464,33 @@ tree, [2]
 
 Thus, `GroupByKey` represents a transform from a multimap (multiple keys to individual values) to a uni-map (unique keys to collections of values).
 
-> **A Note on Key/Value Pairs:** Beam represents key/value pairs slightly differently depending on the language and SDK you're using. In the Beam SDK for Java, you represent a key/value pair with an object of type `KV<K, V>`. In Python, you represent key/value pairs with 2-tuples.
+`CoGroupByKey` is a `PTransform` based on `GroupByKey`, differs from `GroupByKey`, it merges multiple key/value `PCollection`s with the same key type, and emits a collection of `KV<K, CoGbkResult>` pairs. It's usually used as a JOIN operation in Beam pipeline(describe in [design-your-pipeline#multiple-sources]({{ site.baseurl }}/documentation/pipelines/design-your-pipeline/#multiple-sources)).
```
Comment (melap) [inline src/documentation/programming-guide.md]: Your additions would be great as a subsection in Using GroupByKey, so it has a little more visibility. That would help it stand out from the basic GroupByKey text.
*##### Joins with CoGroupByKey*

File: src/documentation/programming-guide.md [hunk 3]
```diff
@@ -464,9 +464,11 @@ tree, [2]
 
 Thus, `GroupByKey` represents a transform from a multimap (multiple keys to individual values) to a uni-map (unique keys to collections of values).
 
-`CoGroupByKey` is a `PTransform` based on `GroupByKey`, differs from `GroupByKey`, it merges multiple key/value `PCollection`s with the same key type, and emits a collection of `KV<K, CoGbkResult>` pairs. It's usually used as a JOIN operation in Beam pipeline(describe in [design-your-pipeline#multiple-sources]({{ site.baseurl }}/documentation/pipelines/design-your-pipeline/#multiple-sources)). 
+##### **Joins with CoGroupByKey**
 
-Given the input collections as below:
+`CoGroupByKey` joins two or more key/value `PCollection`s that have the same key type, and then emits a collection of `KV<K, CoGbkResult>` pairs. [design-your-pipeline#multiple-sources]({{ site.baseurl }}/documentation/pipelines/design-your-pipeline/#multiple-sources) shows an example pipeline that uses a join.
```
Comment (melap) [inline src/documentation/programming-guide.md]: the link text inside the square brackets needs a small change to be Design Your Pipeline. it's not rendering correctly on the [staging site](http://apache-beam-website-pull-requests.storage.googleapis.com/195/documentation/programming-guide/index.html#transforms-gbk):
[Design Your Pipeline]({{ site.baseurl }}/documentation/pipelines/design-your-pipeline/#multiple-sources)

Comment (XuMingmin) [inline src/documentation/programming-guide.md]: updated, thanks @melap

**Comment by XuMingmin** [comment]
Thanks @melap for your review!

**Comment by davorbonaci** [comment]
As a prep for merge, may I ask you to squash and rebase on top of the current `asf-site` branch?

**Comment by XuMingmin** [comment]
rebased.

**Comment by davorbonaci** [comment]
(hm. still conflicts -- will take a look early next week.)

**Comment by XuMingmin** [comment]
let me have a check,,

**Comment by melap** [comment]
@davorbonaci is this mergeable now?

**Comment by dhalperi** [comment]
This has been merged -- thanks @XuMingmin ! Can you please close at your leisure?

**Comment by XuMingmin** [comment]
Thanks @dhalperi !

**Pull request closed without merging**